src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorHtmlPublishDiagnosticsInterceptor.cs

```diff
@@ -0,0 +1,195 @@
+using System.Collections.Generic;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Export(typeof(MessageInterceptor))]
+    [InterceptMethod(Methods.TextDocumentPublishDiagnosticsName)]
+    internal class RazorHtmlPublishDiagnosticsInterceptor : MessageInterceptor
+    {
+        private readonly LSPRequestInvoker _requestInvoker;
+        private readonly LSPDocumentManager _documentManager;
+        private readonly LSPDocumentMappingProvider _documentMappingProvider;
+        private readonly LSPDocumentSynchronizer _documentSynchronizer;
+
+        [ImportingConstructor]
+        public RazorHtmlPublishDiagnosticsInterceptor(
+            LSPRequestInvoker requestInvoker,
+            LSPDocumentManager documentManager,
+            LSPDocumentMappingProvider documentMappingProvider,
+            LSPDocumentSynchronizer documentSynchronizer)
+        {
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            if (documentSynchronizer is null)
+            {
+                throw new ArgumentNullException(nameof(documentSynchronizer));
+            }
+
+            _requestInvoker = requestInvoker;
+            _documentManager = documentManager;
+            _documentMappingProvider = documentMappingProvider;
+            _documentSynchronizer = documentSynchronizer;
+        }
+
+        public override async Task<InterceptionResult> ApplyChangesAsync(JToken token, string containedLanguageName, CancellationToken cancellationToken)
+        {
+            if (token is null)
+            {
+                throw new ArgumentNullException(nameof(token));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            var diagnosticParams = token.ToObject<VSPublishDiagnosticParams>();
+
+            if (diagnosticParams is null)
+            {
+                throw new ArgumentException("Conversion of token failed.");
+            }
+
+            // We only support interception of Virtual HTML Files
+            if (!RazorLSPConventions.IsVirtualHtmlFile(diagnosticParams.Uri))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            var razorDocumentUri = RazorLSPConventions.GetRazorDocumentUri(diagnosticParams.Uri);
+
+            if (!_documentManager.TryGetDocument(razorDocumentUri, out var razorDocumentSnapshot))
+            {
+                return CreateDefaultResponse(token);
```

Comment:
At this point we've failed to resolve the Razor document but we've determined that the filepath that the diagnostics are for looks like a Razor file. Wouldn't a better response be to clear all diagnostics? Otherwise the invalid diagnostics will flow to the platform where they may be presented anyways without any mapping. Same comment with the other failure cases below

Reply:
Makes sense to me, updated to return empty diagnostics in that case.